README
mp3nema
-------

MP3nema is a tool aimed at analyzing and capturing data that is hidden between
frames in an MP3 file or stream, otherwise noted as "out of band" data.  This
utility also supports adding data between frames, and capturing streaming
audio.


Building
--------

From the top-level directory of mp3nema run:
    ./configure
    make

Debugging mode can be enabled when configuring by using the following option:
    ./configure --enable-debug

The resulting binary can be placed anywhere, as there is no "install" target in
the makefile.


Usage
-----
There are four main uses for mp3enema.
0 - Analyzing files or streaming audio to find data hidden between audio frames.
1 - Capturing streaming audio to a file.
2 - Extracting out of band data to a file.
3 - Inserting data between frames.

When inserting data between frames, the destination can be a single MP3 file,
or a directory of MP3s.  To be more covert, larger files should probably be
spanned across multiple MP3s.  In such a case a directory of MP3 files can be
specified.  The resulting files, with the injected data, will be numbered so
that they can be extracted in proper order. 

It is suggested that an ASCII-based encoding (e.g. uuencode) be used to encode
the data that is to be stashed between frames.  This avoids the possibility of
replicating an MP3 sync-frame, which would signify the potential start of audio
data, and confuse the data-extraction process or audio-playback.  Likewise, the
case of an ASCII character sequence of "ID3" in the data might also create
confusion.

The help menu, when running mp3nema without any arguments, designates how to
operate in one of the aforementioned modes.

The files produced are automatically named based on the source MP3 file or
stream that is being analyzed.  In the case of duplicate file names, a
sequential number is added to the end of the resulting file, before the
extension.
    

Thanks
------
The rest of 757labs (www.757labs.com)
GNU (www.gnu.org)


Contact / Project URL
-------
mattdavis9@gmail.com
www.757labs.com/projects/mp3nema


ChangeLog
-- Version 0.4 --
2009-04-07 Matt Davis <mattdavis9@gmail.com>

* main.c: Use a more portable form of exit()
* stream.c: Use a more portable form of exit()


-- Version 0.3 --
2009-04-07 Matt Davis <mattdavis9@gmail.com>

* main.c: Include stdlib.h and updated copyright date
* stream.c: Include arpa/inet.h and netinet/in.h and updated copyright date
* configure.ac: Check for arpa/inet.h and netinet/in.h


-- Version 0.2 --
2008-11-23 Matt Davis <mattdavis9@gmail.com>

* main.c: Added wording to usage statement
* main.h: Up'd the version number
* utils.c: Flush out of band data file pointer when saving out of band data
* utils.c: Always save oob data, even if its just HTTP header info, by ignoring
the ignore flag


-- Version 0.1 --
2008-08-16 Matt Davis <mattdavis9@gmail.com>

* Initial public release 


file.c
/******************************************************************************
 * file.c 
 *
 * mp3nema - MP3 analysis and data hiding utility
 *
 * Copyright (C) 2008 Matt Davis (enferex) of 757Labs (www.757labs.com)
 *
 * file.c is part of mp3nema.
 * mp3nema is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mp3nema is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mp3nema.  If not, see <http://www.gnu.org/licenses/>.
 *****************************************************************************/

#include <stdlib.h>
#include "main.h"
#include "utils.h"


void handle_as_file(const char *fname, flags_t flags)
{
    int            n_frames, n_tags;
    FILE          *fp, *oob_file;
    mp3_frame_t   *frame;
    id3_tag_t     *tag;
    STREAM_OBJECT  type;

    if (!(fp = fopen(fname, "rb")))
      abort();

    oob_file = NULL;
    if (flags & FLAG_EXTRACT_MODE)
      if (!(oob_file = util_create_file(fname, "extracted-oob", "dat", 0)))
        ERR("Could not create a file to store out of band data\n"
            "Normal analysis will still occur.\n");
    
    n_frames = n_tags = 0;

    while ((type = util_next_mp3_frame_or_id3v2(fp, NULL, 0, 0,NULL,oob_file)))
    {
        switch (type)
        {
            case STREAM_OBJECT_MP3_FRAME:
                frame = mp3_get_frame(fp);
                mp3_free_frame(frame);
                ++n_frames;
                break;

            case STREAM_OBJECT_ID3V2_TAG:
                tag = id3_get_tag(fp);
                free(tag);
                ++n_tags;
                break;

            default:
                break;
        }
    }

    printf(TAG " Frames: %d\n", n_frames);
    printf(TAG " ID3v2 Tags: %d\n", n_tags);

    /* Clean */
    if (oob_file)
      fclose(oob_file);
    fclose(fp);
}


main.c
/******************************************************************************
 * main.c 
 *
 * mp3nema - MP3 analysis and data hiding utility
 *
 * Copyright (C) 2009 Matt Davis (enferex) of 757Labs (www.757labs.com)
 *
 * main.c is part of mp3nema.
 * mp3nema is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mp3nema is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mp3nema.  If not, see <http://www.gnu.org/licenses/>.
 *****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "main.h"


flags_t main_flags = 0;


void usage(void)
{
    printf(NAME " Copyright (C) 2009 Matt Davis\n"
           "(enferex) of 757Labs (www.757labs.com)\n"
           "This program comes with ABSOLUTELY NO WARRANTY\n"
           "This is free software, and you are welcome to redistribute it\n"
           "under certain conditions.  For details see the file 'LICENSE'\n"
           "that came with this software or visit:\n"
           "<http://www.gnu.org/licenses/gpl-3.0.txt>\n\n"
           "-- MP3nema v" VERSION " --\n"
           "An MP3 analysis, data capturing, and data hiding utility\n");

    printf("Usage: ./mp3nema <source.mp3 | stream> "
           "[-c] [[-e] | [-i file]] [-v]\n"
           "\t-c Capture audio from network stream\n"
           "\t-i <file> Inject data from 'file' into the mp3 between frames\n"
           "\t-e Extract out of band data to a file\n"
           "\t-v Display more information (out-of-frame data)\n");

    exit(0);
}


int is_file(const char *name)
{
    FILE *fp;

    if ((fp = fopen(name, "rb")))
    {
        fclose(fp);
        return 1;
    }

    return 0;
}


int main(
    int    argc,
    char **argv)
{
    int   i;
    char *datasrc, *fname;

    if (argc < 2)
      usage();

    fname = datasrc = NULL;

    /* Args */
    for (i=1; i<argc; i++)
    {
        /* Insert */
        if (strncmp(argv[i], "-i", 2) == 0)
        {
            if (i+1<argc && argv[i+1][0] != '-')
            {
                datasrc = argv[++i];
                main_flags |= FLAG_INSERT_MODE;
            }
            else
              usage();
        }

        /* Extract to file */
        else if (strncmp(argv[i], "-e", 2) == 0)
          main_flags |= FLAG_EXTRACT_MODE;

        /* Capture Stream */
        else if (strncmp(argv[i], "-c", 2) == 0)
          main_flags |= FLAG_CAPTURE_MODE;

        /* Speak up! */
        else if (strncmp(argv[i], "-v", 2) == 0)
          main_flags |= FLAG_VERBOSE;

        /* Source file or stream */
        else if (argv[i][0] != '-')
          fname = argv[i];
    }

    if (!fname)
      usage();

    if (main_flags & FLAG_INSERT_MODE)
      handle_as_insert(fname, main_flags, datasrc);
    else if (is_file(fname))
      handle_as_file(fname, main_flags);
    else
      handle_as_stream(fname, main_flags);
    
    return 0;
}


utils.c
/******************************************************************************
 * utils.c 
 *
 * mp3nema - MP3 analysis and data hiding utility
 *
 * Copyright (C) 2008 Matt Davis (enferex) of 757Labs (www.757labs.com)
 *
 * utils.c is part of mp3nema.
 * mp3nema is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mp3nema is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mp3nema.  If not, see <http://www.gnu.org/licenses/>.
 *****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "utils.h"


void util_url_to_host_port_file(const char *url, hostdata_t *hostdata)
{
    char *c;

    /* Protocol */
    if (strstr(url, "//"))
      hostdata->host = strdup(strstr(url, "//") + 2);
    else
      hostdata->host = strdup(url);

    /* File location */
    hostdata->file = NULL;
    if ((c = strchr(hostdata->host, '/')) &&
        (*(c+1) != ' ') && (*(c+1) != '\0'))
    {
        hostdata->file = strdup(c);
        *c = '\0';

        if ((c = strchr(hostdata->file, '\r')) ||
            (c = strchr(hostdata->file, '\n')))
          *c = '\0';
    }
    else
      hostdata->file = strdup("/");

    /* Port */
    hostdata->portnum = 80;
    if (strchr(hostdata->host, ':'))
    {
        hostdata->host = strtok(hostdata->host, ":");
        hostdata->port = strdup(strtok(NULL, ":"));
        if ((c = strchr(hostdata->port, '/')))
          *c = '\0';
        hostdata->portnum = atoi(hostdata->port);
    }
    else 
      hostdata->port = strdup("80");
}


FILE *util_create_file(
    const char *fname,
    const char *desc,
    const char *extension,
    int         is_stream)
{
    int          fileno;
    char        *outname;
    const char  *c, *r;
    FILE        *out;
    struct stat  st;

    /* Create output file (in the working directory) */
    if ((c = strrchr(fname, '/')) && strlen(c+1))
      ++c;
    else
      c = fname;

    if ((stat(c, &st) == 0) && S_ISDIR(st.st_mode))
      c = NAME;

    /* Avoid overwriting an existig file by appening a number to the name */
    fileno = 0;
    outname = NULL;
    do 
    {
        free(outname);
        outname = calloc(1, strlen(c) + 64);

        if (is_stream || !(r = strrchr(c, '.')))
          r = c + strlen(c);
    
        strncpy(outname, c, r - c);

        if (fileno > 0)
          sprintf(outname, "%s-%s-%d.%s", outname, desc, fileno, extension);
        else
          sprintf(outname, "%s-%s.%s", outname, desc, extension);
        ++fileno;
    }
    while ((out = fopen(outname, "r")) && !fclose(out));

    if (!(out = fopen(outname, "w")))
    {
        ERR("Could not create output file '%s'\n", outname);
        return NULL;
    }

    free(outname);
    return out;
}


/* Pass either data block or file handle
 * If both are passed, the file handle takes presecendence.
 */
STREAM_OBJECT util_next_mp3_frame_or_id3v2(
    FILE       *fp,
    const char *data,
    int         data_sz,
    int         ignore_oob,
    int        *frame_or_tag_index,
    FILE       *oob_to_file)
{
    int           i, n_blks, oob_size;
    unsigned char v[3] = {0}, *oob;
    long          start, end;
    STREAM_OBJECT ret;

    /* Start OOB data catch fresh */
    oob = NULL;
    n_blks = 0;

    /* Start/end positions for file or indecies into data */ 
    if (fp)
    {
        start = ftell(fp);
        fseek(fp, 0, SEEK_END);
        end = ftell(fp);
        fseek(fp, start, SEEK_SET);
    }
    else if (data)
    {
        start = 0;
        end = data_sz;
    }
    else
      return STREAM_OBJECT_UNKNOWN;

    /* Suck data until we hit another sync frame or id3v2 */
    oob_size = ret = 0;
    while (((start + 3) <= end))
    {
        if (fp && (fread(v, 1, 3, fp) == 0))
          break;
        else if (data)
          memcpy(v, data+start, 3);

        /* Reset start position before we did fread */
        if (fp)
          start = ftell(fp) - 3;

        /* Look for MP3 sync frame */
        if ((v[0] == 0xFF) && ((v[1] & 0xE0) == 0xE0) && 
            mp3_is_valid_frame(fp, start))
        {
            ret = STREAM_OBJECT_MP3_FRAME;
            break;
        }

        /* Not a sync frame, ID3v2 frame? */
        else if ((v[0] == 'I') && (v[1] == 'D') && (v[2] == '3'))
        {
            ret = STREAM_OBJECT_ID3V2_TAG;
            break;
        }

        /* ID3v1 tag (not from a stream) */
        else if (fp && (v[0] == 'T') && (v[1] == 'A') && (v[2] == 'G') &&
                 (start == (end - 128)))
        {
            fseek(fp, 128 - 3, SEEK_CUR);
            continue;
        }

        /* OOB */
        else
        {
            if (oob_size >= (n_blks * OOB_BLK_SIZE))
              oob = realloc(oob, (++n_blks) * OOB_BLK_SIZE);
            oob[oob_size++] = v[0];
        }

        /* Keep lookin */
        if (fp)
          fseek(fp, start + 1, SEEK_SET);
        else
          ++start;
    }

    /* Display OOB data */
    if (oob_size && !ignore_oob && IS_VERBOSE)
    {
        VERBOSE("--OOB Data Found: %d bytes--\n", oob_size);
        for (i=0; i<oob_size; i++)
          VERBOSE("0x%.2x(%c) ", oob[i], 
                 (oob[i] > 31 && oob[i]<127) ? oob[i] : ' ');
        VERBOSE("\n----------------------------\n\n");
    }
    else if (oob_size && !IS_VERBOSE)
      printf(TAG " %d bytes out-of-frame\n", oob_size);

    /* Write OOB data to file */
    if (oob_size && oob_to_file)
    {
        fwrite(oob, oob_size, 1, oob_to_file);
        fflush(oob_to_file);
    }
   
    if (fp)
      fseek(fp, start, SEEK_SET);
    else if (frame_or_tag_index)
      *frame_or_tag_index = start;

    free(oob);
    return ret;
}


mp3_frame_t *mp3_get_frame(FILE *fp)
{
    char         header[6];
    mp3_frame_t *frame;

    if (!(fread(header, 4, 1, fp)))
       return NULL;

    /* Check if certain fields are valid, if not try again */
    frame = calloc(1, sizeof(mp3_frame_t));
    mp3_set_header(frame, header);

    /* Skip CRC data */
    if (frame->crc)
      fseek(fp, 2, SEEK_CUR);

    /* Good header read? */
    if (!mp3_is_valid_header(frame))
    {
        mp3_free_frame(frame);
        return mp3_get_frame(fp);
    }

    if (frame->audio_size < 1)
      return mp3_get_frame(fp);

    /* Suck in the rest of the frame */
    frame->audio = malloc(frame->audio_size);
    fread(frame->audio, frame->audio_size, 1, fp);

    return frame;
}


void mp3_free_frame(mp3_frame_t *frame)
{
    if (!frame)
      return;

    free(frame->audio);
    free(frame);
}


void mp3_write_frame(
    FILE              *fp,
    const mp3_frame_t *frame)
{
    fwrite(frame->header, frame->header_size, 1, fp); 
    fwrite(frame->audio, frame->audio_size, 1, fp); 
}


/* Returns bytes not including frame header
 *
 * Great resource where this information was extracted from
 * http://mpgedit.org/mpgedit/mpeg_format/mpeghdr.htm
 *
 * This site mentions that padding is to be one slot.
 * http://www.codeproject.com/KB/audio-video/mpegaudioinfo.aspx#MPEGAudioFrame
 * In fact, I figured out why my header extraction was improper, after looking
 * at the way they compared their header (sync frame check)
 */
int mp3_frame_length(const mp3_frame_t *frame)
{
    int bit_rate, sample_rate, value_idx;

    /* Bit rate (from header to actual rate value) */
    if ((frame->version == V2) &&
        (((frame->layer == L2)) || (frame->layer == L3)))
      value_idx = 4;
    else if ((frame->version == V2) && (frame->layer == L1))
      value_idx = 3;
    else
      value_idx  = frame->version - frame->layer;

    if ((bit_rate = bitrate_table[frame->bitrate][value_idx] * 1000) < 0)
      return 0;

    /* Sample rate */
    if (frame->version == V1)
      value_idx = 0;
    else if (frame->version == V2)
      value_idx = 1;
    else /* (version == V2_5) */
      value_idx = 2;

    if ((sample_rate = sample_rate_table[frame->samplerate][value_idx]) == 0)
      return 0;
    
    /* Frame length (bytes) */
    if (frame->layer == L1)
      return ((12 * bit_rate / sample_rate + frame->padding) * 4);
    else
      return (144 * bit_rate / sample_rate + frame->padding);
}


/* Sets the values for the fields in the header, leaves audio data NULL */
void mp3_set_header(mp3_frame_t *frame, const char header[4])
{
    memcpy(frame->header, header, 4);
    frame->version = MP3_HDR_VERSION(header);
    frame->layer = MP3_HDR_LAYER(header);
    frame->bitrate = MP3_HDR_BIT_RATE(header);
    frame->samplerate = MP3_HDR_SAMPLE_RATE(header);
    frame->padding = MP3_HDR_PADDING(header);
    frame->crc = MP3_HDR_CRC(header);
    
    if (frame->crc)
      frame->header_size = 6;
    else
      frame->header_size = 4;
    
    /* Set our fp to the beginning of the header */
    frame->audio_size = mp3_frame_length(frame) - frame->header_size;
    frame->audio = NULL;
}


int mp3_is_valid_header(const mp3_frame_t *frame)
{
    return !((frame->samplerate == 0x0003) || (frame->bitrate == 0x000F) ||
            (frame->audio_size < 0));
}


int mp3_is_valid_frame(FILE *fp, long start)
{
    int          valid;
    long         orig;
    char         header[6];
    mp3_frame_t *frame;

    /* TODO Handle stream vs FP data */
    if (!fp)
      return 1;

    orig = ftell(fp);
    fseek(fp, start, SEEK_SET);

    if (!(fread(header, 4, 1, fp)))
    {
        fseek(fp, orig, SEEK_SET);
        return 0;
    }

    frame = calloc(1, sizeof(mp3_frame_t));
    mp3_set_header(frame, header);
    valid = mp3_is_valid_header(frame);

    free(frame);
    fseek(fp, orig, SEEK_SET);

    return valid;
}


id3_tag_t *id3_get_tag(FILE *fp)
{
    id3_tag_t *tag;
    char       header[10];

    fread(header, 10, 1, fp);
    tag = malloc(sizeof(id3_tag_t));
    id3_set_header(tag, header);

    /* Skip the tag data */ 
    fseek(fp, tag->size, SEEK_CUR);

    return tag;
}


void id3_set_header(id3_tag_t *tag, const char header[10])
{    
    tag->size = ID3_HDR_SIZE(header);
    tag->extended_header = ID3_HDR_EXTENDED(header); 
    tag->footer = ID3_HDR_FOOTER(header);
}


utils.h
/******************************************************************************
 * utils.h 
 *
 * mp3nema - MP3 analysis and data hiding utility
 *
 * Copyright (C) 2008 Matt Davis (enferex) of 757Labs (www.757labs.com)
 *
 * utils.h is part of mp3nema.
 * mp3nema is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mp3nema is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mp3nema.  If not, see <http://www.gnu.org/licenses/>.
 *****************************************************************************/
#ifndef UTILS_H_INCLUDE
#define UTILS_H_INCLUDE

#include <stdio.h>
#include "main.h"


typedef struct _hostdata_t 
{
    char *host;
    char *port;
    char *file;
    int   portnum;
} hostdata_t;


/* Returns the host, port, file as strings and the port is also in the returned
 * structure.  This data is extracted from the passed 'url' The populated
 * strings (host, port, file) should be deallocated when through.
 * The hostdata object should already be allocated.
 */
extern void util_url_to_host_port_file(const char *url, hostdata_t *hostdata);


/* Returns a file pointer to a newly created file, in the current working
 * directory, with the description inserted in the name.  If this is a 
 * stream, assume 'fname' is a URL/host, so avoid lopping off an extension,
 * because that would probably be part of the address.
 */
extern FILE *util_create_file(
    const char *fname,
    const char *desc,
    const char *extension,
    int         is_stream);


/* Searches the file stream or data stream for the start of the next mp3 frame
 * or id3v2 tag.  If a data stream is searched, and index into that stream is
 * returned where the frame or tag begins.
 * If 'oob_to_file' is specified, the OOB data is written here.
 */
extern STREAM_OBJECT util_next_mp3_frame_or_id3v2(
    FILE       *fp,
    const char *data,
    int         data_sz,
    int         ignore_oob,
    int        *frame_or_tag_index,
    FILE       *oob_to_file);


/* MP3 Frames */
extern mp3_frame_t *mp3_get_frame(FILE *fp);
extern void mp3_free_frame(mp3_frame_t *frame);
extern void mp3_write_frame(FILE *fp, const mp3_frame_t *frame);
extern int mp3_frame_length(const mp3_frame_t *frame);
extern void mp3_set_header(mp3_frame_t *frame, const char header[4]);
extern int mp3_is_valid_header(const mp3_frame_t *frame);
extern int mp3_is_valid_frame(FILE *fp, long start);


/* ID3 Tags */
extern id3_tag_t *id3_get_tag(FILE *fp);
extern void id3_set_header(id3_tag_t *tag, const char header[10]);


#endif /* UTILS_H_INCLUDE */


main.h
/******************************************************************************
 * main.h 
 *
 * mp3nema - MP3 analysis and data hiding utility
 *
 * Copyright (C) 2008 Matt Davis (enferex) of 757Labs (www.757labs.com)
 *
 * main.h is part of mp3nema.
 * mp3nema is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mp3nema is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mp3nema.  If not, see <http://www.gnu.org/licenses/>.
 *****************************************************************************/

#ifndef MAIN_H_INCLUDE
#define MAIN_H_INCLUDE


/* Thanks to the wonderful resource found at:
 * http://mpgedit.org/mpgedit/mpeg_format/mpeghdr.htm
 */


/* Terminal Display Name */
#define NAME "mp3nema"
#define TAG  "[" NAME "]"

/* Version */
#define _VER(_major, _minor) #_major"."#_minor
#define VERSION _VER(0, 4) /* Major, Minor */

/* Main Argument Flags */
#define FLAG_INSERT_MODE  1
#define FLAG_CAPTURE_MODE 2
#define FLAG_EXTRACT_MODE 4
#define FLAG_VERBOSE      8
typedef unsigned short int flags_t;
extern flags_t main_flags;

/* Error Reporting */
#define ERR(...) {fprintf(stderr, TAG "Error: " __VA_ARGS__);}

/* Verbose */
#define IS_VERBOSE (main_flags & FLAG_VERBOSE)
#define VERBOSE(...) {if (IS_VERBOSE) {printf(__VA_ARGS__);}}

/* For array allocation */
#define DEFAULT_BLK_SZ 512 /* Bytes */

/* Out of Band data (OOB) what we are looking for */
#define OOB_BLK_SIZE DEFAULT_BLK_SZ

/* Number of initial mp3 frames to ignore when injecting mp3 with data */
#define FRAMES_TO_IGNORE 15


typedef struct _mp3_frame_t
{
    unsigned char header[6]; /* 6 bytes 4 for header 2 for CRC if it exists */
    int header_size;
    int version;
    int layer;
    int crc;
    int padding;
    int bitrate;
    int samplerate;
    int audio_size; /* AKA frame length */
    unsigned char *audio;
} mp3_frame_t;


/* What we have found in the mp3 */
typedef enum _stream_thang
{
    STREAM_OBJECT_UNKNOWN,   /* Error */
    STREAM_OBJECT_MP3_FRAME,
    STREAM_OBJECT_ID3V2_TAG
} STREAM_OBJECT;


/* MP3 Version */
#define MP3_HDR_VERSION(_h) ((_h[1] & 0x18) >> 3)
#define V1   0x3
#define V2   0x2
#define V2_5 0x0

/* MP3 Layer */
#define MP3_HDR_LAYER(_h) ((_h[1] & 0x06) >> 1)
#define L1 0x3
#define L2 0x2
#define L3 0x1

/* Other Header Fields */
#define MP3_HDR_CRC(_h)         ((_h[1] & 0x01))
#define MP3_HDR_PADDING(_h)     ((_h[2] & 0x02) >> 1)
#define MP3_HDR_BIT_RATE(_h)    ((_h[2] & 0xF0) >> 4)
#define MP3_HDR_SAMPLE_RATE(_h) ((_h[2] & 0x0C) >> 2)


/* Bit Rate Table
 * 
 * Row index: bitrate value from frame header
 * Col index: Based on MP3 Version and Layer
 */
static const int bitrate_table[][5] = {
    /* 0000 */ {0,  0,  0,  0,  0},
    /* 0001 */ {32, 32, 32, 32, 8},
    /* 0010 */ {64, 48, 40, 48, 16},
    /* 0011 */ {96, 56, 48, 56, 24},
    
    /* 0100 */ {128, 64,  56, 64,  32},
    /* 0101 */ {160, 80,  64, 80,  40},
    /* 0110 */ {192, 96,  80, 96,  48},
    /* 0111 */ {224, 112, 96, 112, 56},

    /* 1000 */ {256, 128, 112, 128, 64},
    /* 1001 */ {288, 160, 128, 144, 80},
    /* 1010 */ {320, 192, 160, 160, 96},
    /* 1011 */ {352, 224, 192, 176, 112},

    /* 1100 */ {384, 256, 224, 192, 128},
    /* 1101 */ {416, 320, 256, 224, 144},
    /* 1110 */ {448, 384, 320, 256, 160},
    /* 1111 */ {-1,  -1,  -1,  -1,  -1},
};


/* Sample rate table
 *
 * Row: index from header
 * Col: index from MP3 version
 */
static const int sample_rate_table[][3] = {
    /* 00 */ {44100, 22050, 11025},
    /* 01 */ {48000, 24000, 12000},
    /* 10 */ {32000, 16000, 8000},
    /* 11 */ {0,     0,     0}
};


/* ID3v2 Tag */
typedef struct _tag
{
    int extended_header;
    int footer;
    unsigned int size;
} id3_tag_t;


#define ID3_HDR_EXTENDED(_h) ((_h[2] & 0x40) >> 4)
#define ID3_HDR_FOOTER(_h)   ((_h[2] & 0x10) >> 4)
#define ID3_HDR_SIZE(_h)  \
    (((_h[6] << 23) | (_h[7] << 16)) | ((_h[8] << 7) | (_h[9])))


/* 
 * Routines
 */

/* Handle the fname as a directory or single mp3 file that data from 'datasrc'
 * file is to be injected into
 */ 
extern void handle_as_insert(
    const char *f_or_dir_name,
    flags_t     flags,
    const char *datasrc);

/* Handle the name as a mp3 file */
extern void handle_as_file(const char *fname, flags_t flags);

/* Only scan the incoming data for OOB info */
extern void handle_as_stream(const char *url, flags_t flags);


#endif /* MAIN_H_INCLUDE */
